### PR TITLE
perf(functional): stop re-deriving static per-call state

### LINKED ISSUE #23291: Functional models rebuild static call state on every forward pass
On every forward pass, `Functional`'s call path redoes work that is fully determined at build time:

- `input_spec` (in `keras/src/models/functional.py`) rebuilds its `InputSpec` list from `self.inputs` on each access, and `_assert_input_compatibility` reads that property twice per call.
- `_run_through_graph` re-sorts the keys of `_nodes_by_depth` every call, even though the graph topology (and thus depth ordering) never changes after the model is built.
- `Functional.call` allocates a fresh `operation_fn` closure per node per call, even for nodes where the closure injects nothing beyond calling the layer directly.

All three are redundant because none of these depend on the input data — they depend only on the static graph structure fixed at build time, so recomputing them per call is pure per-call Python overhead with no runtime information gained.

The fix caches `input_spec`, the sorted depth keys, and the per-node closures once at build time and reads from the cache at call time.

Part of the per-call Python-dispatch overhead series in #22561. Fixed by #23302.

### PR DESCRIPTION
**Pull request by pctablet505**
Closes #23291

## Why this was closed

This change showed no measurable macro benefit on either axis. Its isolated eager effect is within measurement noise on the models tested (all NOISE in the canonical 8-pair paired runs), and it is compile-neutral: 0 graph-break delta, deterministic dynamo decomposition (see the per-PR compile scorecard). It was backed only by a deterministic micro-benchmark (call-count reduction), not a measured end-to-end win, so it does not clear the bar on its own.

torch.compile graph breaks across the #22561 series (this PR does not change the count)

Three pieces of construction-time-fixed state in the `Functional`/`Function` forward path were recomputed on every call: `Functional.input_spec` rebuilt its `InputSpec` list on each property read (and it's read twice per call), `_run_through_graph` re-sorted the static `_nodes_by_depth` keys, and `Functional.call` allocated a fresh per-node `operation_fn` closure even when it had nothing to inject. Each is now computed once at build time and read at call time.

## Why it's safe

- Everything cached is keyed off state that is fixed once the graph is built (`_inputs_struct`, `_nodes_by_depth`). The one mutation path — the manual `input_spec` setter — explicitly invalidates the cache, so a user-supplied spec always wins.
- The `input_spec` cache is written directly to `self.__dict__`, deliberately bypassing `Layer.__setattr__`: routing it through attribute tracking would rewrap the list in a `TrackedList` on every write and break identity stability of the cached object.
- The precomputed depth-key order is read via `getattr` with a re-sort fallback, so instances that never ran `Function.__init__` (old pickles, deserialization edge cases) stay correct — just uncached.
- The closure skip is gated on exactly `training is None and not kwargs`, the only call shape under which the skipped wrapper is provably a no-op (bare `model(x)` inference). Anything with `training` bound or extra kwargs — including `fit()` — takes the unchanged wrapper path.
- Bit-exact vs master: forward max|Δ| = 0.0 on CNN and LLM, `generate(32)` token ids identical.

## Benchmark (GPU, master vs this PR alone)

PR #23302: master vs this PR alone, Keras torch GPU eager, ms

| workload | master | this PR alone |
|---|--:|--:|
| CNN forward | 1.239 ms | 1.230 ms (~noise) |
| LLM forward | 3.520 ms | 3.498 ms (~noise) |
| LLM generate-32 | 113.310 ms | 112.753 ms (~noise) |

8 interleaved A/B process pairs, paired-ratio verdict against an A/A null calibration; torch 2.13.0+cu130, RTX 4050 Laptop GPU, master `a2f8eac8b2`. All three land in noise at the whole-model level — expected, since each removed cost is a tiny fraction of one forward pass. The evidence for this PR is the deterministic elimination of per-call work plus demonstrated no-regression, not a macro win. Note the closure skip only benefits the `model(x)` call shape; the other two items apply to every call.

## Tests

New unit tests in `functional_test.py` and `function_test.py` cover cache correctness and invalidation, closure-skip vs wrapper call shapes, precomputed depth-key use, and fast/slow-path numerical parity. Full `models/`, `ops/function_test.py`, and `layers/layer_test.py` pass on torch; cross-backend sanity runs pass on tensorflow and jax.

Part of the #22561 series.

## Contributor Agreement

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

File: keras/src/models/functional.py [hunk 1]
```diff
@@ -181,12 +181,21 @@ def call(self, inputs, training=None, mask=None, **kwargs):
...
-            ),
-        )
+        if training is None and not kwargs:
+            # No call-context args to inject: `operation_fn(op, **{})`'s
+            # injection loop is a provable no-op here (every candidate value
+            # is `None`), so skip allocating a per-node wrapper closure and
+            # call each operation directly.
+            outputs = self._run_through_graph(
+                inputs, operation_fn=lambda op: op
+            )
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras/src/models/functional.py]: Done in 28163b70eb -- fast path now calls `self._run_through_graph(inputs)` and relies on its own `operation_fn=lambda op: op` default instead of re-passing an identical lambda, avoiding a redundant closure allocation.

File: keras/src/models/functional.py [hunk 2]
```diff
@@ -396,6 +405,16 @@ def input_spec(self):
...
 
+        # Bypass `Layer.__setattr__`/tracking for this cache: it holds a
+        # plain list of `InputSpec` objects, not sublayers/variables, and
+        # going through `self._tracker.track(...)` would rewrap the list
+        # into a new `TrackedList` object on every assignment, which is
+        # both unnecessary bookkeeping and would break identity-stability
+        # of the cached value across reads.
+        cached = self.__dict__.get("_cached_input_spec")
+        if cached is not None:
+            return cached
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras/src/models/functional.py]: Done in 28163b70eb -- input_spec getter now checks `"_cached_input_spec" in self.__dict__` instead of `cached is not None`, so a legitimately-cached `None` (deeply-nested-dict inputs) is honored instead of forcing recomputation on every access.

File: keras/src/models/functional.py [hunk 3]
```diff
@@ -421,16 +440,24 @@ def make_spec_for_tensor(x, name=None):
...
+        # `_inputs_struct` (and therefore the derived spec) is fixed once
+        # the Functional graph is built, so it is safe to cache. The setter
+        # below invalidates the cache if a manual spec is later assigned.
+        self.__dict__["_cached_input_spec"] = spec
+        return spec
 
     @input_spec.setter
     def input_spec(self, value):
         self._manual_input_spec = value
+        self.__dict__["_cached_input_spec"] = None
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras/src/models/functional.py]: Done in 28163b70eb -- input_spec setter now uses `self.__dict__.pop("_cached_input_spec", None)` to actually remove the cache key on invalidation, consistent with the membership-based cache-hit check in the getter.

**Comment by pctablet505** [comment]
Closing this one. The isolated effect is within measurement noise on the models tested — a deterministic micro-optimization (caches construction-time-fixed Functional state) with no measurable end-to-end speedup, and it's compile-neutral. Trimming the #22561 series to the changes with macro impact to keep review focused.
